### PR TITLE
Add ExecutionTime to WorkflowExecutionInfo

### PR DESCRIPTION
**Pull request by alexshtin**
**What changed?**
Add `ExecutionTime` to `WorkflowExecutionInfo`.

**Why?**
Addressing number of TODOs and compute `ExecutionTime` once (when mutable state is create) and reuse it everywhere else. For backward compatibility old code is still there though and should be removed later.

**How did you test it?**
Existing tests.

**Potential risks**
`ExecutionTime` in visibility records used to be `1/1/1970` (epoch 0) for workflows which don't have retry/cron, now it has `ExecutionTime` equal to `StartTime`.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/persistence/visibilityStore.go [hunk 1]
```diff
@@ -273,6 +270,13 @@ func (v *visibilityManagerImpl) convertVisibilityWorkflowExecutionInfo(execution
 		convertedExecution.HistoryLength = execution.HistoryLength
 	}
 
+	// Workflows created before 1.11 have ExecutionTime set to Unix zero time (1/1/1970) for non-cron/non-retry case.
+	// Compute it on the fly using StartTime (if there was a backoff it must be set).
+	// Remove this if block when it is ok to rely on execution.ExecutionTime only (added 6/9/21).
+	if !convertedExecution.ExecutionTime.After(time.Unix(0, 0)) {
+		convertedExecution.ExecutionTime = convertedExecution.StartTime
+	}
+
```
Comment (wxing1292) [inline common/persistence/visibilityStore.go:279]: this new logic looks different that the removed logic? 
```go
// special handling of ExecutionTime for cron or retry	
if execution.ExecutionTime.UnixNano() == 0 {		
    execution.ExecutionTime = execution.StartTime	
}
```

File: service/history/historyEngine.go [hunk 2]
```diff
@@ -1195,15 +1197,20 @@ func (e *historyEngineImpl) DescribeWorkflowExecution(
...
+	if result.GetWorkflowExecutionInfo().GetExecutionTime() == nil {
+		// TODO: we need to consider adding execution time to mutable state
+		// For now execution time will be calculated based on start time and cron schedule/retry policy
+		// each time DescribeWorkflowExecution is called.
+		startEvent, err := mutableState.GetStartEvent()
+		if err != nil {
+			return nil, err
+		}
+		backoffDuration := timestamp.DurationValue(startEvent.GetWorkflowExecutionStartedEventAttributes().GetFirstWorkflowTaskBackoff())
+		result.WorkflowExecutionInfo.ExecutionTime = timestamp.TimePtr(timestamp.TimeValue(result.WorkflowExecutionInfo.GetStartTime()).Add(backoffDuration))
```
Comment (wxing1292) [inline service/history/historyEngine.go:1212]: could you move this logic within mutable state initialization logic here: https://github.com/temporalio/temporal/blob/v1.10.1/service/history/mutableStateBuilder.go#L266

File: service/history/mutableStateBuilder.go [hunk 3]
```diff
@@ -186,6 +186,7 @@ func newMutableStateBuilder(
 	logger log.Logger,
 	namespaceEntry *cache.NamespaceCacheEntry,
 	startTime time.Time,
+	firstWorkflowTaskBackoff *time.Duration,
```
Comment (wxing1292) [inline service/history/mutableStateBuilder.go:189]: `firstWorkflowTaskBackoff *time.Duration` should not be part of the initialization logic

File: service/history/mutableStateBuilder.go [hunk 4]
```diff
@@ -245,6 +246,7 @@ func newMutableStateBuilder(
 		LastProcessedEvent: common.EmptyEventID,
 
 		StartTime:        timestamp.TimePtr(startTime),
+		ExecutionTime:    timestamp.TimePtr(startTime.Add(timestamp.DurationValue(firstWorkflowTaskBackoff))),
```
Comment (wxing1292) [inline service/history/mutableStateBuilder.go:249]: execution time should be overrided by add workflow execution started event

File: service/history/mutableStateBuilder.go [hunk 5]
```diff
@@ -2992,6 +3004,7 @@ func (e *mutableStateBuilder) AddContinueAsNewEvent(
 		e.logger,
 		e.namespaceEntry,
 		timestamp.TimeValue(continueAsNewEvent.GetEventTime()),
+		command.GetBackoffStartInterval(),
```
Comment (wxing1292) [inline service/history/mutableStateBuilder.go:3007]: firstWorkflowTaskBackoff *time.Duration should not be part of the initialization logic

File: service/history/mutableStateBuilder.go (hunk 3 above)
Comment (wxing1292) [reply service/history/mutableStateBuilder.go:189]: the reason start time is part of initialization parameter if for workflow reset case, where start time is the reset time, not the timestamp of start event

File: service/history/mutableStateBuilder.go (hunk 4 above)
Comment (yiminc) [inline service/history/mutableStateBuilder.go:249]: Those times are confusing. I think current StartTime should be named ScheduledTime, and ExecutionTime should be StartTime. :)

**Review by yiminc — APPROVED** [review]

File: service/history/mutableStateBuilder.go (hunk 4 above)
Comment (wxing1292) [reply service/history/mutableStateBuilder.go:249]: agreed

**Review by wxing1292 — CHANGES_REQUESTED** [review]
plz move the backfill logic into one place and remove firstWorkflowTaskBackoff as input parameter of createMutableState

File: service/history/mutableStateBuilder.go (hunk 4 above)
Comment (alexshtin) [reply service/history/mutableStateBuilder.go:249]: Totally agree! We could easily rename them in proto, code, and ES. But they are also mapped to corresponding columns in visibility databases and renaming there is not that easy. So I had to leave it as is.

**Review by wxing1292 — APPROVED** [review]

**Pull request merged**